kv/kvserver: TestStoreRangeMergeRaftSnapshot failed

kv/kvserver.TestStoreRangeMergeRaftSnapshot [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/12024136?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/12024136?buildTab=artifacts#/) on master @ [f3261bf0520beb880593c49394ff1a4c076492cc](https://github.com/cockroachdb/cockroach/commits/f3261bf0520beb880593c49394ff1a4c076492cc):


```
=== RUN   TestStoreRangeMergeRaftSnapshot
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/70ac774d4c136234023a5444c440f71c/logTestStoreRangeMergeRaftSnapshot4024751018
    test_log_scope.go:81: use -show-logs to present logs inline
    client_merge_test.go:3929: 0=auxiliary/sstsnapshot/66/6babf6f0-1f37-4bc3-98fc-40dc9b8e81d3/4.sst
    client_merge_test.go:3940: ssts not byte-wise identical: /artifacts/tmp/_tmp/70ac774d4c136234023a5444c440f71c/ssts1366804908/0_exp.sst and /artifacts/tmp/_tmp/70ac774d4c136234023a5444c440f71c/ssts1366804908/0_act.sst
    client_merge_test.go:4105: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/70ac774d4c136234023a5444c440f71c/logTestStoreRangeMergeRaftSnapshot4024751018
--- FAIL: TestStoreRangeMergeRaftSnapshot (5.19s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStoreRangeMergeRaftSnapshot.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-31992

kv/kvserver.TestStoreRangeMergeRaftSnapshot [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/12025669?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/12025669?buildTab=artifacts#/) on master @ [5d3ba2d23c49820fa91a1f1ab0122ebb684c7cc4](https://github.com/cockroachdb/cockroach/commits/5d3ba2d23c49820fa91a1f1ab0122ebb684c7cc4):


```
=== RUN   TestStoreRangeMergeRaftSnapshot
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/70ac774d4c136234023a5444c440f71c/logTestStoreRangeMergeRaftSnapshot3169797552
    test_log_scope.go:81: use -show-logs to present logs inline
    client_merge_test.go:3929: 0=auxiliary/sstsnapshot/66/b8a25e90-49dc-4a28-abf4-96e46063959d/4.sst
    client_merge_test.go:3940: ssts not byte-wise identical: /artifacts/tmp/_tmp/70ac774d4c136234023a5444c440f71c/ssts1635560061/0_exp.sst and /artifacts/tmp/_tmp/70ac774d4c136234023a5444c440f71c/ssts1635560061/0_act.sst
    client_merge_test.go:4105: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/70ac774d4c136234023a5444c440f71c/logTestStoreRangeMergeRaftSnapshot3169797552
--- FAIL: TestStoreRangeMergeRaftSnapshot (5.20s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStoreRangeMergeRaftSnapshot.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvserver.TestStoreRangeMergeRaftSnapshot [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_UnitTests_BazelUnitTests/12026665?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_UnitTests_BazelUnitTests/12026665?buildTab=artifacts#/) on master @ [4494e945a9070862cfae1b668b3f8b667c41be17](https://github.com/cockroachdb/cockroach/commits/4494e945a9070862cfae1b668b3f8b667c41be17):


```
=== RUN   TestStoreRangeMergeRaftSnapshot
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/7afc027e19e8045a79fb7b548baaab7f/logTestStoreRangeMergeRaftSnapshot1472624995
    test_log_scope.go:81: use -show-logs to present logs inline
    client_merge_test.go:3929: 0=auxiliary/sstsnapshot/66/7a104c25-115d-4c81-8e61-5e72bd6a97c6/4.sst
    client_merge_test.go:3940: ssts not byte-wise identical: /artifacts/tmp/_tmp/7afc027e19e8045a79fb7b548baaab7f/ssts394785567/0_exp.sst and /artifacts/tmp/_tmp/7afc027e19e8045a79fb7b548baaab7f/ssts394785567/0_act.sst
    client_merge_test.go:4105: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/7afc027e19e8045a79fb7b548baaab7f/logTestStoreRangeMergeRaftSnapshot1472624995
--- FAIL: TestStoreRangeMergeRaftSnapshot (5.21s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStoreRangeMergeRaftSnapshot.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvserver.TestStoreRangeMergeRaftSnapshot [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/12027392?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/12027392?buildTab=artifacts#/) on master @ [ef2fd68a7f25fd699d673fc2c4ab5c650979ad16](https://github.com/cockroachdb/cockroach/commits/ef2fd68a7f25fd699d673fc2c4ab5c650979ad16):


```
=== RUN   TestStoreRangeMergeRaftSnapshot
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/70ac774d4c136234023a5444c440f71c/logTestStoreRangeMergeRaftSnapshot3015162620
    test_log_scope.go:81: use -show-logs to present logs inline
    client_merge_test.go:3929: 0=auxiliary/sstsnapshot/66/a8f51d5d-ea43-4a6f-ac44-3815d0a88370/4.sst
    client_merge_test.go:3940: ssts not byte-wise identical: /artifacts/tmp/_tmp/70ac774d4c136234023a5444c440f71c/ssts3234859660/0_exp.sst and /artifacts/tmp/_tmp/70ac774d4c136234023a5444c440f71c/ssts3234859660/0_act.sst
    client_merge_test.go:4105: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/70ac774d4c136234023a5444c440f71c/logTestStoreRangeMergeRaftSnapshot3015162620
--- FAIL: TestStoreRangeMergeRaftSnapshot (5.20s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStoreRangeMergeRaftSnapshot.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvserver.TestStoreRangeMergeRaftSnapshot [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/12028677?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/12028677?buildTab=artifacts#/) on master @ [ef2fd68a7f25fd699d673fc2c4ab5c650979ad16](https://github.com/cockroachdb/cockroach/commits/ef2fd68a7f25fd699d673fc2c4ab5c650979ad16):


```
=== RUN   TestStoreRangeMergeRaftSnapshot
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestStoreRangeMergeRaftSnapshot1497162443
    test_log_scope.go:81: use -show-logs to present logs inline
    client_merge_test.go:3929: 0=auxiliary/sstsnapshot/66/409cc1a6-c90a-44ca-b126-b5dfedf5e6a0/4.sst
    client_merge_test.go:3940: ssts not byte-wise identical: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/ssts2633420000/0_exp.sst and /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/ssts2633420000/0_act.sst
    client_merge_test.go:4105: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestStoreRangeMergeRaftSnapshot1497162443
--- FAIL: TestStoreRangeMergeRaftSnapshot (11.36s)
```
<p>Parameters: <code>TAGS=bazel,gss</code>
, <code>stress=true</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStoreRangeMergeRaftSnapshot.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvserver.TestStoreRangeMergeRaftSnapshot [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/12028635?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/12028635?buildTab=artifacts#/) on master @ [ef2fd68a7f25fd699d673fc2c4ab5c650979ad16](https://github.com/cockroachdb/cockroach/commits/ef2fd68a7f25fd699d673fc2c4ab5c650979ad16):


```
=== RUN   TestStoreRangeMergeRaftSnapshot
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestStoreRangeMergeRaftSnapshot1463822142
    test_log_scope.go:81: use -show-logs to present logs inline
    client_merge_test.go:3929: 0=auxiliary/sstsnapshot/66/be776fe9-6b83-4584-a28d-0afd522d0c6b/4.sst
    client_merge_test.go:3940: ssts not byte-wise identical: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/ssts1087332903/0_exp.sst and /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/ssts1087332903/0_act.sst
    client_merge_test.go:4105: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestStoreRangeMergeRaftSnapshot1463822142
--- FAIL: TestStoreRangeMergeRaftSnapshot (5.25s)
```
<p>Parameters: <code>TAGS=bazel,gss</code>
, <code>stress=true</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStoreRangeMergeRaftSnapshot.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/kv/kvserver/client_merge_test.go
+++ b/pkg/kv/kvserver/client_merge_test.go
@@ -3718,8 +3718,6 @@ func TestStoreRangeMergeRaftSnapshot(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
 
-	skip.UnderMetamorphicWithIssue(t, 111624)
-
 	// We will be testing the SSTs written on store2's engine.
 	var receivingEng, sendingEng storage.Engine
 	// All of these variables will be populated later, after starting the cluster..
@@ -3798,11 +3796,15 @@ func TestStoreRangeMergeRaftSnapshot(t *testing.T) {
 		}
 		keySpans := rditer.MakeReplicatedKeySpans(inSnap.Desc)
 		sstFileWriters := map[string]sstFileWriter{}
-		for _, span := range keySpans {
+		for i, span := range keySpans {
 			file := &storage.MemObject{}
 			writer := storage.MakeIngestionSSTWriter(ctx, st, file)
-			if err := writer.ClearRawRange(span.Key, span.EndKey, true, true); err != nil {
-				return err
+			// If UseExciseForSnapshots is enabled, the snapshot won't write the last
+			// ClearRange in the sstable.
+			if i < len(keySpans)-1 || !storage.UseExciseForSnapshots.Get(&st.SV) {
+				if err := writer.ClearRawRange(span.Key, span.EndKey, true, true); err != nil {
+					return err
+				}
 			}
 			sstFileWriters[string(span.Key)] = sstFileWriter{
 				span:   span,

